packages/financial-templates-lib/src/price-feed/ExpressionPriceFeed.js

```diff
@@ -0,0 +1,117 @@
+const assert = require("assert");
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+const Web3 = require("web3");
+const { create, all } = require("mathjs");
+const math = create(all, {
+  number: "BigNumber",
+  precision: 100
+});
+
+// Gets balancer spot and historical prices. This price feed assumes that it is returning
+// prices as 18 decimals of precision, so it will scale up the pool's price as reported by Balancer contracts
+// if the user specifies that the Balancer contract is returning non-18 decimal precision prices.
+class ExpressionPriceFeed extends PriceFeedInterface {
+  constructor(priceFeedMap, expression) {
+    super();
+    this.expressionCode = math.parse(expression).compile();
+    this.priceFeedMap = priceFeedMap;
+  }
+
+  async getHistoricalPrice(time) {
+    const historicalPrices = {};
+    const errors = [];
+    await Promise.all(
+      Object.entries(this.priceFeedMap).map(async ([name, pf]) => {
+        const price = await pf.getHistoricalPrice(time).catch(err => errors.push(err));
+        historicalPrices[name] = this._convertToDecimal(price);
+      })
+    );
+
+    if (errors.length > 0) {
+      throw errors;
+    }
+
+    return this._convertToFixed(this.expressionCode.evaluate(historicalPrices));
```

Comment:
Does this imply that the `priceFeedMap` has to list pricefeeds in the exact order that they are used in the `expressionCode`?

Reply:
This is a great question. I will add comments as you suggest and we should probably have more checks around detecting if the output is valid. Here's how this roughly works:
```js
let scope = {
  a: 5,
  b: 6
};
let output = math.parse("a + b").compile().evaluate(scope);
```

In this case output is 11. You can use a and b however you like in the expression and they can be any order in the scope. The scope just defines how it resolves each symbol.

However, I'm not sure what happens in this case:
```js
let scope = {
  a: 5
};
let output = math.parse("a + b").compile().evaluate(scope);
```

My guess is that it would return something like "b + 5", but I need to do more testing to make sure we error gracefully in that case and maybe catch it early.

Reply:
It's actually nicer @nicholaspai. It throws if any of the symbols are not resolved at runtime.